orttraining/orttraining/python/orttraining_pybind_state.cc

```diff
@@ -812,42 +824,58 @@ void addObjectMethodsForTraining(py::module& m, ExecutionProviderRegistrationFn
   // It's possible to get the gradient graph through ORTModule by leveraging some "private" fields and not-so-well-documented APIs, so we provide this explicit and tested way to get the gradient graph.
   py::class_<PyGradientGraphBuilder> gradient_graph_builder(m, "GradientGraphBuilder", R"pbdoc(A utility for making a gradient graph that can be used to help train a model.)pbdoc");
   // Set up methods to match the C++ `GradientGraphBuilder` interface.
-  gradient_graph_builder.def(py::init([](
-                                          const py::bytes& serialized_model,
-                                          const std::unordered_set<std::string>& y_node_arg_names,
-                                          const std::unordered_set<std::string>& x_node_arg_names,
-                                          const std::string loss_node_arg_name) {
-                          std::shared_ptr<Model> model;
-                          auto logger_ptr = std::make_unique<logging::Logger>(logging::LoggingManager::DefaultLogger());
-                          logger_ptr->SetSeverity(logging::Severity::kINFO);
-                          ONNX_NAMESPACE::ModelProto model_proto;
-                          std::istringstream model_istream(serialized_model);
-                          ORT_THROW_IF_ERROR(Model::Load(model_istream, &model_proto));
-                          ORT_THROW_IF_ERROR(Model::Load(model_proto, model, nullptr, *logger_ptr));
-                          GradientGraphConfiguration gradient_graph_config{};
-                          gradient_graph_config.set_gradients_as_graph_outputs = true;
-                          // Save some objects, otherwise they get lost.
-                          auto gradient_graph_config_ptr = std::make_unique<GradientGraphConfiguration>(gradient_graph_config);
-
-                          auto builder = std::make_unique<GradientGraphBuilder>(
-                              &model->MainGraph(),
-                              y_node_arg_names,
-                              x_node_arg_names,
-                              loss_node_arg_name,
-                              *gradient_graph_config_ptr,
-                              *logger_ptr);
-
-                          return std::make_unique<PyGradientGraphBuilder>(std::move(builder), std::move(model), std::move(logger_ptr), std::move(gradient_graph_config_ptr));
-                        }))
+  gradient_graph_builder
+      .def(py::init([](const py::bytes& serialized_model,
+                       const std::unordered_set<std::string>& y_node_arg_names,
+                       const std::unordered_set<std::string>& x_node_arg_names,
+                       const std::string loss_node_arg_name,
+                       PySessionOptions* options) {
```

Comment:
I assume this means ORTModule also has the opportunity to support ort-extension later using this one.

Reply:
You're right. This can be extended to be used with ORTModule.
But we can even go further by allowing users to register custom gradient kernels in a custom op library if needed.